driver/vhci_ude/vhci_hc.c

```diff
@@ -84,14 +86,14 @@ setup_vhci(pctx_vhci_t vhci)
 
 	WDF_OBJECT_ATTRIBUTES_INIT(&attrs);
 	attrs.ParentObject = vhci->hdev;
-	status = WdfWaitLockCreate(&attrs, &vhci->lock);
+	status = WdfSpinLockCreate(&attrs, &vhci->spin_lock);
 	if (NT_ERROR(status)) {
-		TRE(VHCI, "failed to create wait lock: %!STATUS!", status);
+		TRE(VHCI, "failed to create spin lock: %!STATUS!", status);
 		return FALSE;
 	}
 	vhci->n_max_ports = MAX_HUB_PORTS;
 
-	vhci->vusbs = ExAllocatePoolWithTag(PagedPool, sizeof(pctx_vusb_t) * vhci->n_max_ports, VHCI_POOLTAG);
+	vhci->vusbs = vhci_pooltag_begin = ExAllocatePoolWithTag(NonPagedPool, sizeof(pctx_vusb_t) * vhci->n_max_ports, VHCI_POOLTAG);
```

Comment:
Yeah, I missed your point. But a vhci cleanup routine can free `vhci->vusbs` instead of using `vhci_pooltag_begin`. Of course, current vhci does not have a cleanup routine. It can be easily added like as a vusb cleanup routine. https://github.com/cezanne/usbip-win/blob/436248caed5c9aaceeac53a6bfa7d5bbbb08e858/driver/vhci_ude/vhci_plugin.c#L206-L218

Reply:
I do no t like my global pointer approach, however i did not know how to do the freeing of 'vusbs' a t that time.

Reply:
cleanup routine for A WDF object of vhci should be added as follows:
 
```
diff --git a/driver/vhci_ude/vhci_hc.c b/driver/vhci_ude/vhci_hc.c
index 813d77a..167343e 100644
--- a/driver/vhci_ude/vhci_hc.c
+++ b/driver/vhci_ude/vhci_hc.c
@@ -126,6 +126,7 @@ evt_add_vhci(_In_ WDFDRIVER drv, _Inout_ PWDFDEVICE_INIT dinit)
        setup_fileobject(dinit);

        WDF_OBJECT_ATTRIBUTES_INIT_CONTEXT_TYPE(&attrs, ctx_vhci_t);
+       attrs.EvtCleanupCallback = vhci_cleanup;
        status = WdfDeviceCreate(&dinit, &attrs, &hdev);
        if (!NT_SUCCESS(status)) {
                TRE(VHCI, "failed to create wdf device: %!STATUS!", status);
```

Reply:
Never mind. I'll try later after your work is merged.